### PR TITLE
Pipes - Iuliia & Kee - Scrabble

### PR DESCRIPTION
**Pull request by anemonekey**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  So we don't have to make a new 'Scoring' instance whenever we want a score.  |
|   Describe an example of a test you wrote for a _nominal case_. |   Checked if won? method would return True when someone accumulates 100 total points.	|
|   Describe an example of a test you wrote for an _edge case_. |   We checked for inputs that were not Strings.	|
|   What was your final test coverage percentage?	|   96.77%	|
|   Is there a specific piece of code you'd like feedback on?	|   We know we have an error with the valid? method in player.rb and we didn't have time to fix it. If there are duplicate letters, we pass them, but we wanted to fail them.	|

### CODE REVIEW

**Comment by Hamled** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  | :white_check_mark: |
|   Both Teammates contributed to the codebase	| :white_check_mark: |
|   Regular Commits with meaningful commit messages  | :white_check_mark: |
|  **Scoring class**	|   	|
|  `score` method	| :white_check_mark: |
|   Uses appropriate data structure to store the letter score	| :white_check_mark: |
|   Appropriately handles edge cases	| Partial. We handle the edge case of an empty or 7-letter word, but do not correctly handle all cases where non-string inputs are provided (e.g. an array or boolean). |
|   Tests `score` edge cases (one-letter word, 7 letter word)	| :white_check_mark: We should add additional tests for non-string inputs, however. |
|  `highest_scoring_word` method	| :white_check_mark: |
|   Appropriately handles edge cases	| :white_check_mark: |
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	| :white_check_mark: We don't technically test for non-array inputs, but otherwise this is a great set of tests covering all of the required functionality! |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	| :white_check_mark: |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	| :white_check_mark: |
|  Returns `true` or `false` from `won?` method | :white_check_mark: |
|  Tests edge cases on the `play(word)` method for words that cannot be scored | No. We should add at least one test to verify the behavior of this code when we attempt to play a word (not Integer or other non-string data type) and it is not valid due to the available letters. |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	| :white_check_mark: |
|  `draw_tiles` method uses the parameter to draw a set of random tiles | :white_check_mark: |
| `tiles_remaining` method returns a count	| :white_check_mark: |
|  Tests `draw_tiles` edge cases for different parameter inputs	| No. We could add tests for providing a non-Integer input, as well as tests for calling `draw_tiles` with zero or too large a value for the `num` parameter. |

File: lib/tile_bag.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    def initialize
+      @tiles = {:A => 9, :B => 2, :C => 2, :D => 4, :E => 12, :F => 2, :G => 3, :H => 2, :I => 9, :J => 1, :K => 1, :L => 4, :M => 2, :N => 6, :O => 8, :P => 2, :Q => 1, :R => 6, :S => 4, :T => 6, :U => 4, :V => 2, :W => 2, :X => 1, :Y => 2, :Z => 1}
+    end #end of initialize
+
+    def draw_tiles(num)
+      available_tiles = @tiles.select { |key, value| value != 0 }
+
+      if tiles_remaining > 0
+        drawn_tiles = available_tiles.keys.sample(num)
```
Comment (Hamled) [inline lib/tile_bag.rb:16]: I think this approach might have a subtle issue that impacts the randomness by which the tiles are drawn.

Specifically, it looks like it only ever draws tiles with distinct letters -- e.g. it would never draw multiple E's from the bag. There's a test that we could write for this which I think would expose the failing behavior (and thus allow you to verify that it has been corrected after making changes):

```ruby
it "should allow drawing duplicate letters" do
  bag = Scrabble::TileBag.new
  drawn_tiles = bag.draw_tiles(27)
  drawn_tiles.length.must_equal 27
end
```

File: lib/tile_bag.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        @tiles.each do |key, value|
+          if drawn_tiles.include?(key.to_sym)
+            @tiles[key] = value - 1
+          end
+        end
+        return drawn_tiles
+      end
+      return "No tiles left"
```
Comment (Hamled) [inline lib/tile_bag.rb:24]: In the case of there being no remaining tiles it would probably be better to return either an empty array or raise an error. That way the code that calls `draw_tiles` can always assume that it will get back an array.

File: specs/player_spec.rb [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    end
+  end
+
+  describe "play(word) method" do
+    it "Does not add empty words to plays Array" do
+      @person.play("")
+      @person.plays.must_be_empty
+    end
+
+    it "Adds words played to plays Array" do
```
Comment (Hamled) [inline specs/player_spec.rb:28]: I really like this test! It's generic enough that if we change unrelated things, like the number of tiles that are drawn by `Player#draw_tiles`, this test will still work and properly test the specific behavior we care about here (updating the `@plays` array).

File: lib/player.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        score = Scoring.score(word)
+        @total_score += score
+
+        index = 0
+        while index < @tiles.length
+          if word.upcase.include?(@tiles[index].to_s)
+            @tiles.delete(@tiles[index])
+          else
+            index += 1
+          end
```
Comment (Hamled) [inline lib/player.rb:50]: This `while` loop code is incorrect. Here is a test case that should expose the issue:

```ruby
it "should remove only the tiles needed to play the given word" do
  # WARNING: This is a hacky thing to force a particular value for the @tiles variable.
  # We would not normally do this as there are much better ways to accomplish the
  # same thing in our tests -- they're just too much code to put into this comment.
  @person.instance_variable_set(:@tiles, ["C", "H", "E", "E", "S", "E", "S"])

  @person.play("CHESS")

  @person.tiles.must_equal ["E", "E"]
end
```

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    end #end of highest_word_score
+
+
+
+    def draw_tiles(tile_bag)
+      if won?
+        return false
+      end
+
+      if tile_bag.tiles_remaining > 0
```
Comment (Hamled) [inline lib/player.rb:81]: All of the logic here around `tiles_remaining` could be eliminated if we made the `TileBag#draw_tiles` method always return an array with a length anywhere from zero up to the number of tiles requested.

Ideally the `Player` class shouldn't have to know so much about the inner workings of `TileBag`.

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    def string?(word)
+      return word.class == String
+    end #end of string?
+
+    def valid?(word)
+      count = 0
+      length = 0
+      for i in 0 ... word.length
+        if @tiles.include?(word[i].upcase.to_sym)
```
Comment (Hamled) [inline lib/player.rb:25]: I think one reason that this `valid?` method is not working properly is that we're not handling the case of duplicate letters in `word`.

In particular, if `@tiles` has the value `["B", "A", "N"]` and we call `valid?("BANANA")` it would return true. This is because we're check each letter in the word independently, when in fact for the last letter we need to check if it's included in `@tiles` _after_ all of the previous letters have been removed from it.

Here's a test which would expose this problem:

```ruby
it "should validate words with duplicate letters" do
  @person.draw_tiles(@tilebag)
  # Sanity check, for this test to work @person.tiles must have more than one distinct letter in it
  @person.tiles.uniq.length.must_be :>, 1

  test_word = @person.tiles.first * 7 # Create a word that is the first tile repeated 7 times
  @person.valid?(test_word).must_equal false
end
```

File: specs/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  describe "valid? method" do
+    it "Will return false when player plays letters not in their hand" do
+      alphabet = @tilebag.tiles.keys.map { |letter| letter.to_s }
+      user_hand = @person.draw_tiles(@tilebag)
+      not_hand = []
+      rand(1..7).times do
+        letter = alphabet.sample(1)
+        not_hand << letter unless user_hand.include? letter
+      end
+      @person.valid?((not_hand).join).must_equal false
```
Comment (Hamled) [inline specs/player_spec.rb:146]: I think this test has a random chance of failing due to the `times` loop above. Here's an example scenario where it might not work:

1. `rand(1..7)` returns 1.
1. `alphabet.sample(1)` returns "A".
1. `user_hand.include? letter` returns true because `@person.tiles` happens to include that letter.
1. We don't push anything into `not_hand` so when we call `valid?(not_hand.join)` we're actually calling it on an empty string.
1. `valid?` as currently written will always return true for an empty string, so this test fails.

What we need to do is find a more consistent way to select a word which contains at least one letter that is not in the player's tile array. One approach to that could be to take `alphabet` and remove all of the letters in `user_hand` from it, leaving us only the letters that were not in `user_hand`.

Also, we should probably have an additional test for `valid?` which verifies that an empty string is not a valid word.

**Review by Hamled — COMMENTED** [review]
Good work on this project!

I especially like how generic your `Player` class tests are! That can get pretty tricky at times, but it's definitely the right kind of thinking to apply when writing your tests.

The more we include "magic numbers" (like expecting a score of 5 for "cat") the more brittle our tests become -- that is, we would need to update the test even when we haven't changed the code that is being tested.
